### PR TITLE
Implement on-the-fly funding based on splicing and liquidity ads

### PR DESCRIPTION
**Pull request by t-bast**
Implement the on-the-fly funding protocol specified in https://github.com/lightning/blips/pull/36: when a payment cannot be relayed because of a liquidity issue, we notify our peer that we'd like to trigger on-the-fly funding if available. If available, we send a funding proposal (`will_add_htlc`) and keep track of its status.

Once a matching funding transaction is signed, we persist this funding attempt and wait for the additional liquidity to be available (once the channel is ready or the splice locked). We will then frequently try to relay the payment to get paid our liquidity fees. If the payment keeps getting rejected, or we cannot connect to our peer, we abandon the
payment when it reaches its CLTV expiry, which ensures that the upstream channels are not at risk.

When using on-the-fly funding, we use a single channel with our peer. If they try to open another channel while one is available, we reject their request and expect a splice instead.

This is best reviewed commit-by-commit: the bulk of the implementation is in the last commit, which contains some subtleties to correctly handle all edge cases (covered by the various unit tests).

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 1]
```diff
@@ -345,6 +491,119 @@ class Peer(val nodeParams: NodeParams,
...
+      expired.foreach {
+        case (paymentHash, pending) => pending.status match {
+          case _: OnTheFlyFunding.Status.Proposed => ()
+          case _: OnTheFlyFunding.Status.Funded => nodeParams.db.liquidity.removePendingOnTheFlyFunding(remoteNodeId, paymentHash)
+        }
+      }
+      pendingOnTheFlyFunding = pendingOnTheFlyFunding.removedAll(expired.keys)
+      d match {
+        case Peer.Nothing => stay()
+        case d if d.channels.isEmpty && pendingOnTheFlyFunding.isEmpty => stopPeer()
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:524]: This is too aggressive for Phoenix: they get constantly disconnected if they don't yet have a channel.

Also, in any case we must explicitly kill the connection when stopping the peer:
```scala
d.peerConnection ! PeerConnection.Kill(KillReason.NoRemainingChannel)
```

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: Yes that's only for the disconnected case, your fix LGTM

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala [hunk 2]
```diff
@@ -161,6 +163,14 @@ object NodeRelay {
       .modify(_.includeLocalChannelCost).setTo(true)
   }
 
+  /** If we fail to relay a payment, we may want to attempt on-the-fly funding if it makes sense. */
+  private def shouldAttemptOnTheFlyFunding(nodeParams: NodeParams, recipientFeatures_opt: Option[Features[InitFeature]], failures: Seq[PaymentFailure]): Boolean = {
+    val featureOk = Features.canUseFeature(nodeParams.features.initFeatures(), recipientFeatures_opt.getOrElse(Features.empty), Features.OnTheFlyFunding)
+    val balanceTooLow = failures.collectFirst { case f@LocalFailure(_, _, BalanceTooLow) => f }.nonEmpty
+    val routeNotFound = failures.collectFirst { case f@LocalFailure(_, _, RouteNotFound) => f }.nonEmpty
+    featureOk && (balanceTooLow || routeNotFound)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala:171]: Both the router and the next channel can return an "insufficient balance" error. It is inherently racy, but always happens in the context of splice outs due to #2912.

Also some additional logs help for debugging.

```suggestion
    val routerBalanceTooLow = failures.collectFirst { case f@LocalFailure(_, _, BalanceTooLow) => f }.nonEmpty
    val channelBalanceTooLow = failures.collectFirst { case f@LocalFailure(_, _, _: InsufficientFunds) => f }.nonEmpty
    val routeNotFound = failures.collectFirst { case f@LocalFailure(_, _, RouteNotFound) => f }.nonEmpty
    val res = featureOk && (routerBalanceTooLow || channelBalanceTooLow || routeNotFound)
    context.log.info(s"on-the-fly funding assessment: shouldAttemptPayToOpen=$res featureOk=$featureOk routerBalanceTooLow=$routerBalanceTooLow channelBalanceTooLow=$channelBalanceTooLow localFailures={}", failures.collect { case f: LocalFailure => f.t.getClass.getSimpleName }.mkString(","))
    res
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala [hunk 3]
```diff
@@ -306,14 +316,14 @@ class NodeRelay private(nodeParams: NodeParams,
...
+          relay(upstream, recipient, Some(r.remoteFeatures), nextPayload, nextPacket_opt)
       }
     }
   }
 
   /** Relay the payment to the next identified node: this is similar to sending an outgoing payment. */
-  private def relay(upstream: Upstream.Hot.Trampoline, recipient: Recipient, payloadOut: IntermediatePayload.NodeRelay, packetOut_opt: Option[OnionRoutingPacket]): Behavior[Command] = {
+  private def relay(upstream: Upstream.Hot.Trampoline, recipient: Recipient, recipientFeatures_opt: Option[Features[InitFeature]], payloadOut: IntermediatePayload.NodeRelay, packetOut_opt: Option[OnionRoutingPacket]): Behavior[Command] = {
     context.log.debug("relaying trampoline payment (amountIn={} expiryIn={} amountOut={} expiryOut={})", upstream.amountIn, upstream.expiryIn, payloadOut.amountToForward, payloadOut.outgoingCltv)
     val confidence = (upstream.received.map(_.add.endorsement).min + 0.5) / 8
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala:328]: We only need one attempt when the next node is a wallet. This reduces diff with feature branches

```suggestion
  private def relay(upstream: Upstream.Hot.Trampoline, recipient: Recipient, walletNodeId_opt: Option[PublicKey], recipientFeatures_opt: Option[Features[InitFeature]], payloadOut: IntermediatePayload.NodeRelay, packetOut_opt: Option[OnionRoutingPacket]): Behavior[Command] = {
    context.log.debug("relaying trampoline payment (amountIn={} expiryIn={} amountOut={} expiryOut={})", upstream.amountIn, upstream.expiryIn, payloadOut.amountToForward, payloadOut.outgoingCltv)
    val confidence = (upstream.received.map(_.add.endorsement).min + 0.5) / 8
    // We only make one try when it's a direct payment to a wallet
    val maxPaymentAttempts = if (walletNodeId_opt.isDefined) 1 else nodeParams.maxPaymentAttempts
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/OnTheFlyFunding.scala [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+    private def waitForResult(remaining: Int): Behavior[Command] = {
+      Behaviors.receiveMessagePartial {
+        case WrappedCommandResponse(response) => response match {
+          case _: CommandSuccess[_] => Behaviors.same
+          case failure: CommandFailure[_, _] =>
+            cmd.replyTo ! RelayFailed(paymentHash, CannotAddToChannel(failure.t))
+            stopping(remaining - 1)
+        }
+        case WrappedHtlcSettled(settled) =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/OnTheFlyFunding.scala:266]: I find the parallel processing of `WrappedCommandResponse`/`WrappedHtlcSettled` a little bit hard to follow, tried a stash-based alternative in 00b1efa5d.

**Review by pm47 — APPROVED** [review]
